### ISSUE TITLE
xconv for segmentation problem -- implementation and memory efficiency

### ISSUE THREAD

**Issue by lodurality** [pyg-team/pytorch_geometric#1470]
## ❓ Questions & Help

Hi, I have implemented PointCNN version for segmentation using PyTorch Geometric but my version is less memory efficient than TensorFlow one. Here is my code, including modified XConv (to use upsampling). It also shows subpar results on PartNet dataset which might be due to small batch size and related batch normalization issues. 

Here is my modified XConv module. I used your advice in the issue #821 and added knn computation instead of knn_graph but I wanted to double-check with you that I got it correct. If it is indeed correct, I can commit it as contribution to issue #821 

```
from typing import Optional

from math import ceil

import torch
import torch.nn.functional as F
from torch import Tensor
from torch.nn import Sequential as Seq, Linear as Lin, BatchNorm1d as BN
from torch.nn import ELU, Conv1d, Dropout
from torch_geometric.nn import Reshape
from torch_geometric.nn import fps, global_mean_pool

try:
    from torch_cluster import knn_graph, knn
except ImportError:
    knn_graph = None


def reset(nn):
    def _reset(item):
        if hasattr(item, 'reset_parameters'):
            item.reset_parameters()

    if nn is not None:
        if hasattr(nn, 'children') and len(list(nn.children())) > 0:
            for item in nn.children():
                _reset(item)
        else:
            _reset(nn)


class XConv(torch.nn.Module):
    def __init__(self, in_channels: int, out_channels: int, dim: int,
                 kernel_size: int, hidden_channels: Optional[int] = None,
                 dilation: int = 1, bias: bool = True, num_workers: int = 1,
                 with_global: bool = False, bn_momentum: float = 0.5):
        super(XConv, self).__init__()

        if knn_graph is None:
            raise ImportError('`XConv` requires `torch-cluster`.')

        # print(with_global)
        self.in_channels = in_channels
        if hidden_channels is None:
            hidden_channels = in_channels // 4
        assert hidden_channels > 0
        self.hidden_channels = hidden_channels
        self.out_channels = out_channels
        self.dim = dim
        self.kernel_size = kernel_size
        self.dilation = dilation
        self.num_workers = num_workers
        self.with_global = with_global

        C_in, C_delta, C_out = in_channels, hidden_channels, out_channels
        D, K = dim, kernel_size

        self.mlp1 = Seq(
            Lin(dim, C_delta),
            ELU(),
            BN(C_delta, momentum=bn_momentum),
            Lin(C_delta, C_delta),
            ELU(),
            BN(C_delta, momentum=bn_momentum),
            Reshape(-1, K, C_delta),
        )

        self.mlp2 = Seq(
            Lin(D * K, K ** 2),
            ELU(),
            BN(K ** 2, momentum=bn_momentum),
            Reshape(-1, K, K),
            Conv1d(K, K ** 2, K, groups=K),
            ELU(),
            BN(K ** 2, momentum=bn_momentum),
            Reshape(-1, K, K),
            Conv1d(K, K ** 2, K, groups=K),
            BN(K ** 2, momentum=bn_momentum),
            Reshape(-1, K, K),
        )

        if with_global:
            self.global_mlp = Seq(Lin(3, out_channels // 4, bias=True),
                                  ELU(),
                                  BN(out_channels // 4, momentum=bn_momentum),
                                  Lin(out_channels // 4, out_channels // 4, bias=True),
                                  ELU(),
                                  BN(out_channels // 4, momentum=bn_momentum),
                                  )

        C_in = C_in + C_delta
        depth_multiplier = int(ceil(C_out / C_in))
        self.conv = Seq(
            Conv1d(C_in, C_in * depth_multiplier, K, groups=C_in),
            Reshape(-1, C_in * depth_multiplier),
            Lin(C_in * depth_multiplier, C_out, bias=bias),
        )

        self.reset_parameters()

    def reset_parameters(self):
        reset(self.mlp1)
        torch.nn.init.xavier_normal(self.mlp1[0].weight)
        torch.nn.init.xavier_normal(self.mlp1[3].weight)
        reset(self.mlp2)
        torch.nn.init.xavier_normal(self.mlp2[0].weight)
        torch.nn.init.xavier_normal(self.mlp2[4].weight)
        torch.nn.init.xavier_normal(self.mlp2[8].weight)
        reset(self.conv)

    def forward(self, x: Tensor, pos: Tensor, batch: Optional[Tensor] = None,
                pos_query: Optional[Tensor] = None, batch_query: Optional[Tensor] = None):
        """"""
        pos = pos.unsqueeze(-1) if pos.dim() == 1 else pos
        pos_backup = pos.clone()
        (N, D), K = pos.size(), self.kernel_size

        if pos_query is None:
            edge_index = knn_graph(pos, K * self.dilation, batch, loop=True,
                                   flow='target_to_source',
                                   num_workers=self.num_workers)
            row, col = edge_index[0], edge_index[1]
        else:
            # print('deconv')
            edge_index = knn(pos_query, pos, K * self.dilation, batch_x=batch_query, batch_y=batch)
            row, col = edge_index[0], edge_index[1]
            # print(row, col)
            # print(row.max(), col.max(), row.shape, col.shape)

        if self.dilation > 1:
            dil = self.dilation
            index = torch.randint(K * dil, (N, K), dtype=torch.long,
                                  device=row.device)
            arange = torch.arange(N, dtype=torch.long, device=row.device)
            arange = arange * (K * dil)
            index = (index + arange.view(-1, 1)).view(-1)

            # print(row, index)
            row = row[index]
            col = col[index]

        pos = pos[col] - pos[row]

        x_star = self.mlp1(pos.view(N * K, D))
        if x is not None:
            x = x.unsqueeze(-1) if x.dim() == 1 else x
            x = x[col].view(N, K, self.in_channels)
            x_star = torch.cat([x_star, x], dim=-1)
        x_star = x_star.transpose(1, 2).contiguous()
        x_star = x_star.view(N, self.in_channels + self.hidden_channels, K, 1)

        transform_matrix = self.mlp2(pos.view(N, K * D))
        transform_matrix = transform_matrix.view(N, 1, K, K)

        x_transformed = torch.matmul(transform_matrix, x_star)
        x_transformed = x_transformed.view(N, -1, K)

        out = self.conv(x_transformed)
        # print(out.shape)

        if self.with_global:
            # print(pos.shape)
            fts_global = self.global_mlp(pos_backup)
            # print(fts_global.shape)
            return torch.cat([fts_global, out], axis=-1)
        else:
            return out

    def __repr__(self):
        return '{}({}, {})'.format(self.__class__.__name__, self.in_channels,
                                   self.out_channels)
```

And here is my model -- I don't see here obvious memory bottlenecks. 

```
class PointCNN(torch.nn.Module):
    def __init__(self, num_classes, bn_momentum=0.01):
        super(PointCNN, self).__init__()

        self.num_classes = num_classes

        self.conv1 = XConv(0, 256, dim=3, kernel_size=8, hidden_channels=256 // 2, dilation=1)
        self.conv2 = XConv(256, 256, dim=3, kernel_size=12, hidden_channels=256 // 4, dilation=2)
        self.conv3 = XConv(256, 512, dim=3, kernel_size=16, hidden_channels=512 // 4, dilation=2)
        self.conv4 = XConv(512, 1024, dim=3, kernel_size=16, hidden_channels=1024 // 4, dilation=6, with_global=True)

        self.deconv1 = XConv(1024 + 1024 // 4, 1024, dim=3, kernel_size=16, hidden_channels=512 // 4, dilation=6)
        self.deconv2 = XConv(1024, 512, dim=3, kernel_size=16, hidden_channels=256 // 4, dilation=6)
        self.deconv3 = XConv(512, 256, dim=3, kernel_size=12, hidden_channels=256 // 4, dilation=6)
        self.deconv4 = XConv(256, 256, dim=3, kernel_size=8, hidden_channels=256 // 4, dilation=6)
        self.deconv5 = XConv(256, 256, dim=3, kernel_size=8, hidden_channels=256 // 4, dilation=4)

        self.fuse1 = Seq(Lin(2048 + 1024 // 4, 1024, bias=True), ELU(), BN(1024, momentum=bn_momentum))
        self.fuse2 = Seq(Lin(1024, 512, bias=True), ELU(), BN(512, momentum=bn_momentum))
        self.fuse3 = Seq(Lin(512, 256, bias=True), ELU(), BN(256, momentum=bn_momentum))
        self.fuse4 = Seq(Lin(512, 256, bias=True), ELU(), BN(256, momentum=bn_momentum))
        self.fuse5 = Seq(Lin(512, 256, bias=True), ELU(), BN(256, momentum=bn_momentum))

        self.lin1 = Seq(Lin(256, 256, bias=True), ELU(), BN(256, momentum=bn_momentum))
        torch.nn.init.xavier_uniform(self.lin1[0].weight)
        self.lin2 = Seq(Lin(256, 256, bias=True), ELU(), BN(256, momentum=bn_momentum))
        torch.nn.init.xavier_uniform(self.lin1[0].weight)
        self.lin3 = Lin(256, num_classes)
        torch.nn.init.xavier_uniform(self.lin1[0].weight)

    def forward(self, x, pos, batch, edge_index=None):
        x1 = F.relu(self.conv1(None, pos, batch))
        
        idx1 = fps(pos, batch, ratio=0.375)
        x1_sub, pos1_sub, batch1_sub = x1[idx1], pos[idx1], batch[idx1]

        x2 = F.relu(self.conv2(x1_sub, pos1_sub, batch1_sub))

        idx2 = fps(pos1_sub, batch1_sub, ratio=0.5)
        x2_sub, pos2_sub, batch2_sub = x2[idx2], pos1_sub[idx2], batch1_sub[idx2]

        x3 = F.relu(self.conv3(x2_sub, pos2_sub, batch2_sub))

        idx3 = fps(pos2_sub, batch2_sub, ratio=1 / 3)
        x3_sub, pos3_sub, batch3_sub = x3[idx3], pos2_sub[idx3], batch2_sub[idx3]
        x4 = F.relu(self.conv4(x3_sub, pos3_sub, batch3_sub))
        x = F.relu(self.deconv1(x4, pos3_sub, batch3_sub, pos_query=pos3_sub, batch_query=batch3_sub))
        x = torch.cat([x, x4], axis=-1)

        x = self.fuse1(x)
        x = F.relu(self.deconv2(x, pos2_sub, batch2_sub, pos_query=pos3_sub, batch_query=batch3_sub))

        x = torch.cat([x, x3], axis=-1)
        x = self.fuse2(x)
        
        x = F.relu(self.deconv3(x, pos1_sub, batch1_sub, pos_query=pos2_sub, batch_query=batch2_sub))

        x = torch.cat([x, x2], axis=-1)
        x = self.fuse3(x)
        
        x = F.relu(self.deconv4(x, pos, batch, pos_query=pos1_sub, batch_query=batch1_sub))

        x = torch.cat([x, x1], axis=-1)
        x = self.fuse4(x)
        
        x = F.relu(self.deconv5(x, pos, batch))

        x = torch.cat([x, x1], axis=-1)
        x = self.fuse5(x)

        x = F.relu(self.lin1(x))
        x = F.relu(self.lin2(x))
        x = F.dropout(x, p=0.5, training=self.training)
        x = self.lin3(x)
        return F.log_softmax(x, dim=-1)

```

Thanks for your amazing work!

**Comment by lodurality** [comment]
Maybe the issue is that PointCNN uses C++ compiled sampling which uses less memory but it is a wild guess.

**Comment by rusty1s** [comment]
That's looks quite amazing so far. However, I must admit that I don't have enough experience with the `XConv` segmentation architecture to spot the differences to the official implementation. 

In general, I see no reason why this implementation should consume more memory than the official implementation. Do you have any insights into the "C++ compiled sampling"? Why should it consume less memory?

Looking forward to your progress and a PR if you have time :)

**Comment by Crane-YU** [comment]
@rusty1s sorry, one quick question, for the dimension of pos, is D the dimension of the positions, say 3 in 3D space. Thank you.

**Comment by rusty1s** [comment]
Yes :)

**Comment by Crane-YU** [comment]
@rusty1s thank you

**Comment by Crane-YU** [comment]
@lodurality Hi, could you plz explain why you use 2 non-linear functions right before and after batchnorm?

**Comment by lodurality** [comment]
@Crane-YU what part of the code do you refer to (can't find it)?

**Comment by Crane-YU** [comment]
@lodurality so basically u define the layers as
`self.lin1 = Seq(Lin(256, 256, bias=True), ELU(), BN(256, momentum=bn_momentum))
self.lin2 = Seq(Lin(256, 256, bias=True), ELU(), BN(256, momentum=bn_momentum))
`
then in the last few lines of ur code u have:
`x = F.relu(self.lin1(x))
x = F.relu(self.lin2(x))`

**Comment by lodurality** [comment]
@Crane-YU good catch, thanks! I will fix it.

**Comment by edshkim98** [comment]
@lodurality Hi, I have a question about your code.
x4 = F.relu(self.conv4(x3_sub, pos3_sub, batch3_sub))
        x = F.relu(self.deconv1(x4, pos3_sub, batch3_sub, pos_query=pos3_sub, batch_query=batch3_sub))
        x = torch.cat([x, x4], axis=-1)
In this part, shouldnt the input channel of deconv1 be 1024 which is same as out channel of conv4 in order to be able to make it work?

**Comment by edshkim98** [comment]
@lodurality Also, how did you set your hidden channel as 256//2 ,512//4 and so on. I couldnt find any information on the paper about setting hidden channel for part segmentation?

**Comment by edshkim98** [comment]
@lodurality Hi, have you managed to solve the memory/performance issue?

**Comment by ryangoy** [comment]
Hi, I thought I'd leave a comment here in case anyone else has this issue. 

In the original PointCNN code, each input point cloud and label is sampled to 2048 points before it is fed into the model. Most of the bottleneck in what @lodurality  wrote is in the first layer since it is not sampled (I tested with a custom dataset with 100k points in each input). 

However, in any case, it is better to use the original PointCNN code if you are interested in speed and efficiency. I ported the original code to Tensorflow 2.4 and it runs several times faster to what I got pytorch to run.

**Comment by Bear-kai** [comment]
#2601

**Comment by tashfin-msu** [comment]
@ryangoy 
Hi. By any chance, have you made your pointcnn tensorflow 2.4 code public?
Thanks

**Comment by nnMan1** [comment]
Hi. It can be because in official implementation, sampling is performed before xconv operation, so convolution is done only in query points, not in all point. And in encoder part of the network xconv operation is porform over all points, and after that you take fps points. In any case yours work is amazing.
